### ISSUE TITLE
Siemens S7, stop running around in circles.

### ISSUE THREAD

**Issue by magnumripper** [magnumripper/JohnTheRipper#1098]
We've been too quick to make uneducated hacks in the Siemens format lately. I have the feeling the current canonicalization (change *0* to *1*) is somewhat flawed. We should probably strip that field altogether in the canonical .pot representation.

**BUT** my bottom line is *we should sort it out* before we do anything now.

**Comment by frank-dittrich** [comment]
@magnum, you are right, thanks for digging into this.

I should have included the run directory in my `git grep siemsns-s7`, so that I would have noticed that the difference between '0' and '1' is on purpose.
Arguably, the '0' and '1' should be in some other field (uid/gid), so that you could limit your password search to specific hashes (if that makes sense).

About the reverting my split() implementation for siemens-s7.
I did run the self test and a real test run against old and new test hashes, but not with ./jtrts.pl.
When I realized that the TS didn't pass, I tested wether reverting my last commit would fix it, and it looked like I it worked before.
Only later I noticed that the TS test hashes had an issue, and that valid() would still accept these invalid hashes.

I think the best action now is to change one self test back to using '0', and adding comments to run/pcap2john.py and src/mssql12_fmt_plug.c about the meaning of '0' and '1'.

**Comment by magnumripper** [comment]
Yes, that self-test is already reverted to original. Regarding the rest, we need to verify, not guess. My **guess** is that `*0*` means "unsuccessful auth" which means we can still crack the password but it wasn't correct for that system (it might be valuable for some other system?). But again, we can't keep guessing. We do not add any such comments based on guesses.
